Improve Python API docs

kianfar77 opened the pull request:
## What changes are proposed in this pull request?
This PR replaces #154 with more improvements on the documentation of Python API functions including better organization, explanation, formatting, and addition of a "returns" field to the function docs.  

## How is this patch tested?
- [ ] Unit tests
- [x] Integration tests
- [x] Manual tests

(Details)

karenfeng commented on functions.yml:
```diff
@@ -187,26 +201,23 @@ etl:
       doc: |
-             Normalize the variant with a behavior similar to vt normalize or bcftools norm.
+             Normalizes the variant with a behavior similar to vt normalize or bcftools norm.
              Creates a StructType column including the normalized start, end, referenceAllele and
```
It may make more sense to move this extensive description to its own section in the variant normalization doc page

kianfar77 replied:
Left it in as we talked. Will address var normalization doc in a followup PR

karenfeng commented on functions.yml:
```diff
@@ -231,15 +242,15 @@ etl:
           doc: |
-            A path to the reference genome .fasta file. The .fasta file must
-            be accompanied with a .fai index file in the same folder.
+            A path to the reference genome .fasta file. The .fasta file must be accompanied with a .fai index file in the same folder.
```
Wrap the file types in backticks

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -187,26 +201,23 @@ etl:
 
-             changed: A boolean field indicating whether the variant data was changed as a
-                 result of normalization.
+                changed: A boolean field indicating whether the variant data was changed as a result of normalization.
```
Wrap `changed` in backticks.

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -187,26 +201,23 @@ etl:
-             errorMessage: An error message in case the attempt at normalizing the row hit an
-                 error. In this case, the changed field will be set to false. If no errors occur,
-                 this field will be null.
+                errorMessage: An error message in case the attempt at normalizing the row hit an error. In this case, the changed field will be set to false. If no errors occur, this field will be null.
```
Wrap `errorMessage` in backticks. We probably don't need to wrap generic types, but we should wrap any values (eg. `false`, `null`)

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -187,26 +201,23 @@ etl:
 
-             In case of an error, the start, end, referenceAllele and alternateAlleles fields in
-             the generated struct will be null.
+             In case of an error, the start, end, referenceAllele and alternateAlleles fields in the generated struct will be null.
```
Wrap `start`, `end`, `referenceAllele`, `alternateAlleles`, and `null` in backticks.

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -407,16 +429,17 @@ gwas_functions:
+          doc: An array of genotype structs
         - name: phenotypes
-          doc: An array of phenotype values
+          doc: A ``double`` array of phenotype values
```
double -> floating-point

kianfar77 replied:
removed wrapping

karenfeng commented on docs/source/api-docs/toplevel-functions.rst:
```diff
@@ -0,0 +1,5 @@
+Glow Top-level Functions
```
`Top-level` -> `Top-Level`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -77,8 +81,9 @@ complex_type_manipulation:
+      returns: A ``spark.ml`` ``SparseVector``
     - name: expand_struct
-      doc: Promote fields of a nested struct to top-level columns. Similar to using struct.* from SQL, but can be used in more contexts.
+      doc: Promotes fields of a nested struct to top-level columns. Similar to using struct.* from SQL, but can be used in more contexts.
```
Wrap `struct.*` in backticks to clarify that it's SQL code

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -89,8 +94,9 @@ complex_type_manipulation:
+      returns: Columns corresponding to fields of the input struct
     - name: explode_matrix
-      doc: Explode a spark.ml Matrix into arrays of rows
+      doc: Explodes a ``spark.ml`` matrix (sparse or dense) into multiple arrays, one per row of the matrix.
```
For consistency, we should write out `spark.ml` `Matrix`; we don't need to specify that it's sparse/dense as those are the only two implementation (`SparseMatrix` and `DenseMatrix`)

kianfar77 replied:
changed to ``Matrix``. Left sparse or dense for emphasis

karenfeng commented on functions.yml:
```diff
@@ -271,8 +283,9 @@ quality_control:
+      returns: A struct containing ``callRate``, ``nCalled``, ``nUncalled``, ``nHet``, ``nHomozygous``, ``nNonRef``, ``nAllelesCalled``, ``alleleCounts``, ``alleleFrequencies`` fields. See :ref:`variant-qc`.
     - name: dp_summary_stats
-      doc: Compute summary statistics for depth field from array of genotype structs
+      doc: Compute summary statistics for depth field from array of genotype structs. See :ref:`variant-qc`.
```
Compute -> Computes

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -103,8 +109,9 @@ complex_type_manipulation:
       args:
         - name: matrix
           doc: The matrix to explode
```
The `spark.ml` `Matrix` to explode

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -132,11 +140,13 @@ complex_type_manipulation:
   functions:
     - name: hard_calls
-      doc: Converts an array of probabilities to hard calls
+      doc: Converts an array of probabilities to hard calls. See :ref:`variant-data-transformations` 
```
Converts an array of probabilities to hard calls. The probabilites are assumed to be diploid. See :ref:`variant-data-transformations`.

kianfar77 replied:
Changed

karenfeng commented on functions.yml:
```diff
@@ -153,17 +163,21 @@ etl:
+          doc: The number of alternate alleles
         - name: phased
-          doc: Whether the probabilities are phased or not
+          doc: Whether the probabilities are phased. If phased, we expect one 2 * numAlts values 
```
Wrap any mathematical expressions in backticks

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -153,17 +163,21 @@ etl:
         - name: phased
-          doc: Whether the probabilities are phased or not
+          doc: Whether the probabilities are phased. If phased, we expect one 2 * numAlts values 
+            in probabilities array. If unphased, we expect one probability per possible genotype.
```
`in probabilities array` -> `in the probabilities array`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -153,17 +163,21 @@ etl:
         - name: threshold
-          doc: The minimum probability to include
+          doc: The minimum probability to make a call. If no probability falls into the range 
+            of [0, 1 - threshold] or [threshold, 1], a no-call (represented by -1s) will be emitted. 
```
Wrap mathematical expressions in backticks

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -153,17 +163,21 @@ etl:
-          doc: The minimum probability to include
+          doc: The minimum probability to make a call. If no probability falls into the range 
+            of [0, 1 - threshold] or [threshold, 1], a no-call (represented by -1s) will be emitted. 
+            If not provided, this parameter defaults to 0.9.
```
Wrap `0.9`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -187,26 +201,23 @@ etl:
           type: str
         - name: minMatchRatio
-          doc: Minimum fraction of bases that must remap to lift over successfully
+          doc: Minimum fraction of bases that must remap to do liftover successfully. If not provided, defaults to 0.95.
```
Wrap `0.95` in backticks

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -389,9 +409,11 @@ gwas_functions:
           doc: An array of phenotypes
         - name: covariates
-          doc: A Spark matrix of covariates
+          doc: A spark.ml matrix of covariates
```
`spark.ml` `Matrix`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -389,9 +409,11 @@ gwas_functions:
         - name: phenotypes
           doc: An array of phenotypes
```
A numeric array of phenotypes

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -335,6 +351,7 @@ quality_control:
           doc: An array of alternate alleles
+      returns: A struct containing ``sampleId``, ``callRate``, ``nCalled``, ``nUncalled``, ``nHomRef``, ``nHet``, ``nHomVar``, ``nSnp``, ``nInsertion``, ``nDeletion``, ``nTransition``, ``nTransversion``, ``nSpanningDeletion``, ``rTiTv``, ``rInsertionDeletion``, ``rHetHomVar`` fields. See :ref:`sample-qc`.
     - name: sample_dp_summary_stats
       doc: Compute per-sample summary statistics about the depth field in an array of genotype structs
```
`Compute per-sample summary statistics about the depth field in an array of genotype structs` -> `Computes per-sample summary statistics about the depth field in an array of genotype structs.`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -350,7 +367,8 @@ quality_control:
+          doc: An array of genotype structs
+      returns: An array of structs where each struct contains ``mean``, ``stDev``, ``min``, and ``max`` of the genotype depths for a sample. If ``sampleId`` is present in a genotype, it will be propagated to the resulting struct as an extra field.
     - name: sample_gq_summary_stats
       doc: Compute per-sample summary statistics about the genotype quality field in an array of genotype structs
```
`Compute per-sample summary statistics about the genotype quality field in an array of genotype structs` -> `Computes per-sample summary statistics about the genotype quality field in an array of genotype structs.`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -335,6 +351,7 @@ quality_control:
           doc: The reference allele
```
For two lines above: `The array of genotype structs` -> `An array of genotype structs with a `calls` field`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -350,7 +367,8 @@ quality_control:
       args:
         - name: genotypes
-          doc: The array of genotype structs
+          doc: An array of genotype structs
```
`An array of genotype structs` -> `An array of genotype structs with a ``depth`` field`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -366,12 +384,14 @@ quality_control:
       args:
         - name: genotypes
-          doc: The array of genotype structs
+          doc: An array of genotype structs with a conditionalQuality field
```
Wrap `conditionalQuality` in backticks

kianfar77 replied:
Done

kianfar77 commented on functions.yml:
```diff
@@ -35,12 +35,14 @@ complex_type_manipulation:
+      returns: A struct consisting of the input struct and the added fields
     - name: array_summary_stats
-      doc: Compute the min, max, mean, stddev for an array of numerics
+      doc: Computes the min, max, mean, stddev for an array of numerics. The input array can be any numeric
```
Changed

kianfar77 commented on functions.yml:
```diff
@@ -35,12 +35,14 @@ complex_type_manipulation:
     - name: array_summary_stats
-      doc: Compute the min, max, mean, stddev for an array of numerics
+      doc: Computes the min, max, mean, stddev for an array of numerics. The input array can be any numeric
+        type, but the returned stats are always ``double`` typed.
```
Moved the sentence to return. As to ``float`` vs ``double``, I think since these are short for DoubleType os Spark double makes more sense.

kianfar77 commented on functions.yml:
```diff
@@ -51,8 +53,9 @@ complex_type_manipulation:
       args:
         - name: arr
           doc: The array of numerics
+      returns: A struct containing ``mean``, ``stdDev``, ``min``, and ``max`` fields
```
Changed

kianfar77 commented on functions.yml:
```diff
@@ -153,17 +163,21 @@ etl:
+      returns: An array of hard calls
     - name: lift_over_coordinates
-      doc: Do liftover like Picard
+      doc: Performs liftover for the coordinates of a variants. To perform liftover of alleles and add additional metadata, see :ref:`liftover`.
```
Done

kianfar77 commented on functions.yml:
```diff
@@ -407,16 +429,17 @@ gwas_functions:
+      returns: A struct containing ``beta``, ``oddsRatio``, ``waldConfidenceInterval``, and ``pValue`` fields. See :ref:`logistic-regression`.
     - name: genotype_states
-      doc: Get number of alt alleles for a genotype
+      doc: Gets number of alt alleles for genotypes. Returns ``-1`` if there are any ``-1``s in the calls array.
```
Done

kianfar77 commented on functions.yml:
```diff
@@ -407,16 +429,17 @@ gwas_functions:
+          doc: A ``double`` array of phenotype values
         - name: covariates
-          doc: a matrix of covariates
+          doc: M ``spark.ml`` matrix of covariates
```
Done

karenfeng commented on functions.yml:
```diff
@@ -432,3 +454,4 @@ gwas_functions:
       args:
         - name: genotypes
           doc: An array of genotype structs
```
An array of genotype structs with ``calls`` field

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -407,16 +428,17 @@ gwas_functions:
+      returns: A struct containing ``beta``, ``oddsRatio``, ``waldConfidenceInterval``, and ``pValue`` fields. See :ref:`logistic-regression`.
     - name: genotype_states
-      doc: Get number of alt alleles for a genotype
+      doc: Gets number of alternate alleles for genotypes. Returns ``-1`` if there are any ``-1`` s in the calls array.
```
`Gets number of alternate alleles for genotypes.` -> `Gets the number of alternate alleles for an array of genotype structs.` We should also clarify that `-1` represents a missing call.

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -315,8 +329,9 @@ quality_control:
       args:
         - name: genotypes
           doc: The array of genotype structs
```
An array of genotype structs with `conditionalQuality` field

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -153,17 +162,21 @@ etl:
         - name: threshold
-          doc: The minimum probability to include
+          doc: The minimum probability to make a call. If no probability falls into the range 
+            of ``[0, 1 - threshold]`` or ``[threshold, 1]``, a no-call (represented by ``-1``s) will be emitted.
```
The `-1` didn't render properly because it's immediately followed by an `s`

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -299,8 +312,9 @@ quality_control:
       args:
         - name: genotypes
           doc: The array of genotype structs
```
An array of genotype structs with `calls` field

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -187,26 +200,23 @@ etl:
           doc: Location of the chain file on each node in the cluster
```
For above: replace the doc for `The current contigName` -> `The current contig name`

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -387,11 +406,13 @@ gwas_functions:
         - name: genotypes
           doc: An array of genotypes
```
A numeric array of genotypes (to distinguish for the user from genotype structs)

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -407,16 +428,17 @@ gwas_functions:
       args:
         - name: genotypes
-          doc: An array of genotypes
+          doc: An array of genotype structs
```
A numeric array of genotypes

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -187,26 +200,23 @@ etl:
-             errorMessage: An error message in case the attempt at normalizing the row hit an
-                 error. In this case, the changed field will be set to false. If no errors occur,
-                 this field will be null.
+                ``errorMessage``: An error message in case the attempt at normalizing the row hit an error. In this case, the changed field will be set to false. If no errors occur, this field will be null.
```
Wrap `false` and `null` in backticks

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -231,15 +241,15 @@ etl:
           doc: Alternate alleles
```
For above: replace `The current contigName` with `The current contig name`. Replace `Reference allele` with `The reference allele`. Here, replace `Alternate alleles` with `An array of alternate alleles`

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -366,12 +383,14 @@ quality_control:
       args:
         - name: genotypes
-          doc: The array of genotype structs
+          doc: An array of genotype structs with a ``conditionalQuality`` field
```
We should be consistent with whether we preface the field name with `a` or not

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -116,11 +122,12 @@ complex_type_manipulation:
+      returns: A struct containing only the indicated fields
     - name: vector_to_array
-      doc: Convert a spark.ml vector (sparse or dense) to an array of doubles
+      doc: Converts a ``spark.ml`` vector (sparse or dense) to an array of doubles.
```
`spark.ml.linalg` `Vector`

kianfar77 replied:
Fixed

karenfeng commented on functions.yml:
```diff
@@ -50,9 +51,10 @@ complex_type_manipulation:
         - name: arr
-          doc: The array of numerics
+          doc: An array of any numeric type
+      returns: A struct containing double-typed ``mean``, ``stdDev``, ``min``, and ``max`` fields.
```
double-typed -> double, for consistency

kianfar77 replied:
changed

kianfar77 commented on functions.yml:
```diff
@@ -271,8 +282,9 @@ quality_control:
+      returns: A struct containing ``callRate``, ``nCalled``, ``nUncalled``, ``nHet``, ``nHomozygous``, ``nNonRef``, ``nAllelesCalled``, ``alleleCounts``, ``alleleFrequencies`` fields. See :ref:`variant-qc`.
     - name: dp_summary_stats
-      doc: Compute summary statistics for depth field from array of genotype structs
+      doc: Computes summary statistics for depth field from array of genotype structs. See :ref:`variant-qc`.
```
Done

kianfar77 commented:
> Looking really good! Just to nit-pick, can we be consistent on whether the docs/parameters/return values end with a period or not? Respectively, I think it's traditional to go yes/no/no.

Addressed the comments. Double checked the periods following your suggestion except for when a complete sentence is used for arg docs and return where I put a period.

karenfeng commented on functions.yml:
```diff
@@ -271,8 +282,9 @@ quality_control:
+      returns: A struct containing ``callRate``, ``nCalled``, ``nUncalled``, ``nHet``, ``nHomozygous``, ``nNonRef``, ``nAllelesCalled``, ``alleleCounts``, ``alleleFrequencies`` fields. See :ref:`variant-qc`.
     - name: dp_summary_stats
-      doc: Compute summary statistics for depth field from array of genotype structs
+      doc: Computes summary statistics for the depth field from array of genotype structs. See :ref:`variant-qc`.
```
from array -> from an array

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -407,16 +428,17 @@ gwas_functions:
+      returns: A struct containing ``beta``, ``oddsRatio``, ``waldConfidenceInterval``, and ``pValue`` fields. See :ref:`logistic-regression`.
     - name: genotype_states
-      doc: Get number of alt alleles for a genotype
+      doc: Gets number of alternate alleles for an array of genotype structs. Returns ``-1`` if there are any ``-1`` s (no-calls) in the calls array.
```
Gets number -> Gets the number

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -298,9 +311,10 @@ quality_control:
       args:
         - name: genotypes
-          doc: The array of genotype structs
+          doc: The array of genotype structs with ``calls`` field``
```
Remove trailing ``

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -283,8 +295,9 @@ quality_control:
       args:
         - name: genotypes
           doc: The array of genotype structs
```
An array of genotype structs with `depth` field

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -271,8 +282,9 @@ quality_control:
       args:
         - name: genotypes
           doc: The array of genotype structs
```
An array of genotype structs with `calls` field

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -116,11 +122,12 @@ complex_type_manipulation:
+      returns: A struct containing only the indicated fields
     - name: vector_to_array
-      doc: Convert a spark.ml vector (sparse or dense) to an array of doubles
+      doc: Converts a ``spark.ml.linalg`` ``Vector`` (sparse or dense) to an array of doubles.
```
Oh sorry, didn't realize we just haven't been writing out `linalg`. Can you remove that again? Sorry.

kianfar77 replied:
Done

karenfeng commented on functions.yml:
```diff
@@ -407,16 +428,17 @@ gwas_functions:
-          doc: a matrix of covariates
+          doc: A ``spark.ml`` ``Matrix`` of covariates
         - name: test
           doc: Which logistic regression test to use. Can be 'LRT' or 'Firth'
```
Can we wrap `'LRT'` and `'Firth'` in backticks?

kianfar77 replied:
Done
